cli/src/cmd/forge/fourbyte.rs

```diff
@@ -0,0 +1,119 @@
+use crate::{
+    cmd::forge::build::{CoreBuildArgs, ProjectPathsArgs},
+    compile,
+    opts::forge::CompilerArgs,
+};
+use clap::Parser;
+use ethers::prelude::artifacts::{output_selection::ContractOutputSelection, LosslessAbi};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use tracing::trace;
+
+#[derive(Serialize, Debug)]
+struct ImportRequest {
+    #[serde(rename = "type")]
+    import_type: String,
+    data: Vec<LosslessAbi>,
+}
+
+#[derive(Deserialize, Debug)]
+struct ImportTypeData {
+    imported: HashMap<String, String>,
+    duplicated: HashMap<String, String>,
+}
+
+#[derive(Deserialize, Debug)]
+struct ImportData {
+    function: ImportTypeData,
+    event: ImportTypeData,
+}
+
+#[derive(Deserialize, Debug)]
+struct ImportResponse {
+    result: ImportData,
+}
+
+#[derive(Debug, Clone, Parser)]
+pub struct UploadSelectorsArgs {
+    #[clap(help = "The name of the contract to upload selectors for.")]
+    pub contract: String,
+
+    #[clap(flatten, next_help_heading = "PROJECT OPTIONS")]
+    pub project_paths: ProjectPathsArgs,
+}
+
+impl UploadSelectorsArgs {
+    /// Builds a contract and uploads the ABI to selector database
+    pub async fn run(self) -> eyre::Result<()> {
+        let UploadSelectorsArgs { contract, project_paths } = self;
+
+        let build_args = CoreBuildArgs {
+            project_paths: project_paths.clone(),
+            out_path: Default::default(),
+            ignored_error_codes: vec![],
+            no_auto_detect: false,
+            use_solc: None,
+            offline: false,
+            force: false,
+            libraries: vec![],
+            via_ir: false,
+            revert_strings: None,
+            compiler: CompilerArgs {
+                extra_output: vec![ContractOutputSelection::Abi],
+                ..Default::default()
+            },
+        };
+
+        trace!("Building project");
+        let project = build_args.project()?;
+        let outcome = compile::suppress_compile(&project)?;
+        let found_artifact = outcome.find(&contract);
+        let artifact = found_artifact.ok_or_else(|| {
+            eyre::eyre!("Could not find artifact `{contract}` in the compiled artifacts")
+        })?;
+
+        let body = ImportRequest {
+            import_type: "abi".to_string(),
+            data: vec![artifact.abi.clone().ok_or(eyre::eyre!("Unable to fetch abi"))?],
```

Comment:
can we `take` here to get rid of the clone?

Reply:
Hmmm - probably lack of rust skill, but attempts to do this didn't quite work because `arbifact.abi` is a reference and can't be borrowed as mutable (required by `artifact.abi.take()`). Unaware of a way to take a reference out without copy or clone - would love a suggestion